base_geo_route_optimization/models/geo_route_optimize.py

```diff
@@ -0,0 +1,189 @@
+# Copyright (C) 2020 Brian McMaster
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+import logging
+import json
+import urllib
+
+from ortools.constraint_solver import routing_enums_pb2
+from ortools.constraint_solver import pywrapcp
+
+from odoo import _, api, models
+from odoo.exceptions import ValidationError
+
+_logger = logging.getLogger(__name__)
+
+
+class GeoRouteOptimize(models.AbstractModel):
+    ''' This model is meant to solve the Traveling Salesman Problem
+       and/or the Vehicle Routing Problem.
+       It is meant to be inherited by any model that needs to efficiently
+       optimize the sequence in which records are ordered, specifically for
+       driving purposes.
+       Distances between records are calculated and stored in a matrix that is
+       generated by the Google Distance Matrix API
+       The optimization problem is solved using Google OR Tools python package
+       The functions in this model are based on the examples provided in the
+       following link.
+       More information at https://developers.google.com/optimization/routing
+    '''
+    _name = 'geo.route.optimize'
+    _description = 'Geo Route Optimization Framework'
+
+    def send_request(self, origin_addresses, dest_addresses):
+        ''' Build and send the API request for the given origin and
+           destination addresses.
+        '''
+        def build_address_str(addresses):
+            # Build a pipe-separated string of addresses
+            address_str = ''
+            for i in range(len(addresses) - 1):
+                address_str += addresses[i] + '|'
+            address_str += addresses[-1]
+            return address_str
```

Comment:
You might just want to use `join` here. Or am I missing something?

```suggestion
            return '|'.join(addresses)
```

Reply:
Good call. Updated